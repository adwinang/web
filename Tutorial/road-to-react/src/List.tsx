import { memo, useId, useState } from "react";
import { ListProps, Stories, Story } from "./type";
import styles from "./App.module.css";
import { ReactComponent as Check } from "./check.svg";
import { sortBy } from "lodash";

interface ISort {
  NONE(list: Stories): Stories;
  TITLE(list: Stories): Stories;
  AUTHOR(list: Stories): Stories;
  COMMENT(list: Stories): Stories;
  POINT(list: Stories): Stories;
}

type SortType = keyof ISort;

const SORTS: ISort = {
  NONE: (list: Stories): Stories => list,
  TITLE: (list: Stories): Stories => sortBy(list, "title"),
  AUTHOR: (list: Stories): Stories => sortBy(list, "author"),
  COMMENT: (list: Stories): Stories => sortBy(list, "num_comments").reverse(),
  POINT: (list: Stories): Stories => sortBy(list, "points").reverse(),
};

const Item = ({
  item,
  onRemoveItem,
}: {
  item: Story;
  onRemoveItem: (item: Story) => void;
}) => {
  // Components only accept one input
  return (
    <li className={styles.item}>
      <span style={{ width: "40%" }}>
        <a href={item.url}>{item.title}</a>
      </span>
      <span style={{ width: "30%" }}> {item.author} </span>
      <span style={{ width: "10%" }}>{item.num_comments} </span>
      <span style={{ width: "10%" }}>{item.points} </span>
      <span style={{ width: "10%" }}>
        {/* 
        Alternative implementation for onClick  
        onClick={onRemoveItem.bind(null, item)}
        Though I would not recommend it as it would require everyone to under "this" scoping
        Current implementation have a caveat which is that we should avoid any complex logic in the arrow function
        As the JavaScript logic may be hidden in JSX
        */}
        <button
          type="button"
          onClick={() => onRemoveItem(item)}
          className={`${styles.button} ${styles.button_small}`}
        >
          <Check style={{ height: "18px", width: "18px" }} />
        </button>
      </span>
    </li>
  );
};

const List = memo(({ list, onRemoveItem }: ListProps) => {
  const id = useId();
  const [sort, setSort] = useState<{ sortKey: SortType; isReverse: Boolean }>({
    sortKey: "NONE" as SortType,
    isReverse: false,
  });

  const handleSort = (sortKey: SortType) => {
    const isReverse = sort.sortKey === sortKey && !sort.isReverse;
    setSort({ sortKey, isReverse });
  };

  const sortFunction = SORTS[sort.sortKey];
  const sortedList = sort.isReverse
    ? sortFunction(list).reverse()
    : sortFunction(list);

  return (
    <ul>
      <li className={styles.item}>
        <span style={{ width: "40%" }}>
          <button type="button" onClick={() => handleSort("TITLE" as SortType)}>
            <b>Title</b>
          </button>
        </span>
        <span style={{ width: "30%" }}>
          <button
            type="button"
            onClick={() => handleSort("AUTHOR" as SortType)}
          >
            <b>Author</b>
          </button>
        </span>
        <span style={{ width: "10%" }}>
          <button
            type="button"
            onClick={() => handleSort("COMMENT" as SortType)}
          >
            <b>Comments</b>
          </button>
        </span>
        <span style={{ width: "10%" }}>
          <button type="button" onClick={() => handleSort("POINT" as SortType)}>
            <b>Points</b>
          </button>
        </span>
        <span style={{ width: "10%" }}>
          <b>Actions</b>
        </span>
      </li>
      {sortedList.map((item) => {
        return (
          <Item
            key={`${id}-${item.objectID}`}
            onRemoveItem={onRemoveItem}
            item={item}
          />
        );
      })}
    </ul>
  );
});

export { List, Item };
