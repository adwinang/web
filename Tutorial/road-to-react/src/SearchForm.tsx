import styles from "./App.module.css";
import { InputWithLabel } from "./InputWithLabel";
import { SearchFormProps } from "./type";

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit,
}: SearchFormProps) => (
  <form onSubmit={onSearchSubmit} className={styles.searchForm}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      isFocused
      onInputChange={onSearchInput}
    >
      <strong>Search:</strong>
    </InputWithLabel>
    <button
      type="submit"
      disabled={!searchTerm}
      className={`${styles.button} ${styles.buttonLarge}`}
    >
      Submit
    </button>
  </form>
);

export { SearchForm };
