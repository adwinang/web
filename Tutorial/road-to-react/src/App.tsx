import { LastSearches } from "./LastSearches";
import {
  ChangeEvent,
  FormEvent,
  useCallback,
  useEffect,
  useReducer,
  useRef,
  useState,
} from "react";
import axios from "axios";
import styled from "styled-components";
import { StoriesAction, StoriesState, Story } from "./type";
import { List } from "./List";
import { SearchForm } from "./SearchForm";
const API_BASE = "https://hn.algolia.com/api/v1";
const API_SEARCH = "/search";
const PARAM_SEARCH = "query=";
const PARAM_PAGE = "page=";

const StyledContainer = styled.div`
  padding: 20px;
  background: #83a4d4;
  background: linear-gradient(to left, #b6fbff, #83a4d4);
  color: #171212;
`;
const StyledHeadlinePrimary = styled.h1`
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
`;

const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case "STORIES_FETCH_INIT":
      return { ...state, isLoading: true, isError: false };
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data:
          action.payload.page === 0
            ? action.payload.list
            : state.data.concat(action.payload.list),
        page: action.payload.page,
      };
    case "STORIES_FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

const useSemiPersistentState = (
  key: string,
  initialState: string
): [string, (value: string) => void] => {
  // If false, useEffect only runs once, if true, useEffect only runs on update
  const isMounted = useRef(false);
  const [value, setValue] = useState(localStorage.getItem(key) || initialState);
  // First argument is the function that runs the side-effect.
  // Second Argument is a dependency array of variables
  // If any of the depended variables is changed, the function is ran
  // Leaving out the second argument would make the function
  // for the side effect run on every render (initial render and update renders) of the component.
  // The setting of items is deemed as a side effect
  // useEffect is a side-effect hook
  useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = !isMounted.current;
    } else {
      localStorage.setItem(key, value);
    }
  }, [value, key]);

  return [value, setValue];
};

const getSumComments = (stories: StoriesState) => {
  return stories.data.reduce((result, value) => result + value.num_comments, 0);
};

const getUrl = (searchTerm: string, page: number): string =>
  `${API_BASE}${API_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`;

const extractSearchTerm = (url: string): string =>
  url
    .substring(url.lastIndexOf("?") + 1, url.lastIndexOf("&"))
    .replace(PARAM_SEARCH, "");

const getLastSearches = (urls: Array<string>) =>
  urls.reduce((result: Array<string>, url, index) => {
    const searchTerm = extractSearchTerm(url);
    if (index === 0) {
      return result.concat(searchTerm);
    }
    const previousSearchTerm = result[result.length - 1];
    if (searchTerm === previousSearchTerm) {
      return result;
    } else {
      return result.concat(searchTerm);
    }
  }, []);

const App = () => {
  const [stories, dispatchStories] = useReducer(storiesReducer, {
    data: [],
    page: 0,
    isLoading: false,
    isError: false,
  });
  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
  const [urls, setUrls] = useState<Array<string>>([getUrl(searchTerm, 0)]);

  // useCallback is one of Reacts memo hooks
  // useCallback only creates a memoized function every time its dependency array changes.
  // Only then will the useEffect hook run again because it depends on the new function
  const handleFetchStories = useCallback(async () => {
    dispatchStories({ type: "STORIES_FETCH_INIT", page: 0 });
    try {
      const lastUrl = urls[urls.length - 1];
      const result = await axios.get(lastUrl);

      console.log(result);

      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: { list: result.data.hits, page: result.data.page },
      });
    } catch (err) {
      dispatchStories({ type: "STORIES_FETCH_FAILURE" });
    }
  }, [urls]);

  useEffect(() => {
    handleFetchStories();
    // The Empty array will tell useEffect to only run once the component renders for the first time and it will not run again
  }, [handleFetchStories]);

  // Tells React not to re-render the function
  // Should only use if the function has a dependency on states within the component
  // In this case there is not, hence there is no use for the useCallback below
  const handleRemoveStory = useCallback((item: Story) => {
    dispatchStories({ type: "REMOVE_STORY", payload: item });
  }, []);

  const handleSearch = (searchTerm: string, page: number) => {
    const url = getUrl(searchTerm, page);
    setUrls(urls.concat(url));
  };

  const handleSearchInput = (event: ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event: FormEvent<HTMLFormElement>) => {
    handleSearch(searchTerm, 0);
    event.preventDefault();
  };

  const handleLastSearch = (searchTerm: string) => {
    setSearchTerm(searchTerm);
    handleSearch(searchTerm, 0);
  };

  const handleMore = () => {
    const lastUrl = urls[urls.length - 1];
    const searchTerm = extractSearchTerm(lastUrl);
    handleSearch(searchTerm, stories.page + 1);
  };

  const lastSearches = getLastSearches(urls);

  const sumComments = getSumComments(stories);

  return (
    <StyledContainer>
      <StyledHeadlinePrimary>
        My Hacker Stories with {sumComments} comments
      </StyledHeadlinePrimary>
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />
      <LastSearches
        lastSearches={lastSearches}
        onLastSearch={handleLastSearch}
      />
      {/*In JavaScript a true && "Hello World" always evaluates to "Hellow
      World" Whereas a false && "Hello World" always evaluates to false*/}

      {stories.isError && <p>Something went wrong ...</p>}

      {stories.isLoading ? (
        <p>Loading ...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}

      <button type="button" onClick={handleMore}>
        More
      </button>
    </StyledContainer>
  );
};

export default App;
export { storiesReducer };
