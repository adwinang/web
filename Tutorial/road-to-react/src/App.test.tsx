import TestRenderer from "react-test-renderer";
import { render, screen, fireEvent, act } from "@testing-library/react";
import axios from "axios";
import App, { storiesReducer } from "./App";
import { StoriesRemoveAction, StoriesState, Story } from "./type";
import { Item } from "./List";
import { SearchForm } from "./SearchForm";

jest.mock("axios");
const mockedAxios = jest.mocked(axios, true);

const storyOne: Story = {
  title: "React",
  url: "https://reactjs.org/",
  author: "Jordan Walke",
  num_comments: 3,
  points: 4,
  objectID: "0",
};

const storyTwo = {
  title: "Redux",
  url: "https://redux.js.org/",
  author: "Dan Abramov, Andrew Clark",
  num_comments: 2,
  points: 5,
  objectID: 1,
};
const stories = [storyOne, storyTwo];

describe("storiesReducer", () => {
  test("removes a story from all stories", () => {
    const action: StoriesRemoveAction = {
      type: "REMOVE_STORY",
      payload: storyOne,
    };
    const state: StoriesState = {
      data: stories,
      isLoading: false,
      isError: false,
      page: 0,
    };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [storyTwo],
      isLoading: false,
      isError: false,
    };

    // If use toBe, there will be a problem as the object reference between the two objects are not the same
    expect(newState).toStrictEqual(expectedState);
  });
});

describe("Item", () => {
  test("renders all properties", () => {
    render(
      <Item
        item={storyOne}
        onRemoveItem={() => {
          return;
        }}
      />
    );
    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();
    expect(screen.getByText("React")).toHaveAttribute(
      "href",
      "https://reactjs.org/"
    );
  });

  test("renders a clickable dismiss button", () => {
    render(
      <Item
        item={storyOne}
        onRemoveItem={() => {
          return;
        }}
      />
    );
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("clicking the dismiss button calls the callback handler", () => {
    // The following function is a mock function to capture when it's called
    const handleRemoveItem = jest.fn();
    render(<Item item={storyOne} onRemoveItem={handleRemoveItem} />);

    fireEvent.click(screen.getByRole("button"));

    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
  });
});

describe("SearchForm", () => {
  const searchFormProps = {
    searchTerm: "React",
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn(),
  };
  test("renders the input field with its value", () => {
    render(<SearchForm {...searchFormProps} />);
    expect(screen.getByDisplayValue("React")).toBeInTheDocument();
  });

  test("renders the correct label", () => {
    render(<SearchForm {...searchFormProps} />);
    // Regular Expressions can be used as well
    expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
  });

  test("calls onSearchInput on input field change", () => {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.change(screen.getByDisplayValue("React"), {
      target: { value: "Redux" },
    });
    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
  });

  test("calls onSearchSubmit on button submit click", () => {
    render(<SearchForm {...searchFormProps} />);
    fireEvent.submit(screen.getByRole("button"));
    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
  });

  test("renders snapshot", () => {
    // React Test Renderer library is recommended for testing snapshots
    const search = TestRenderer.create(
      <SearchForm {...searchFormProps} />
    ).toJSON();
    expect(search).toMatchSnapshot();
  });
});

describe("App component", () => {
  test("succeeds fetching data", async () => {
    // Simulate fetching of data
    const promise: Promise<any> = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    mockedAxios.get.mockImplementationOnce(() => promise);

    render(<App />);

    // There is a difference between getByText and queryByText
    // getByText will throw an error if the element is not found
    // queryByText will return null if the element is not found which may not be intended
    expect(screen.getByText(/Loading/)).toBeInTheDocument();

    await act(async () => promise);

    expect(screen.queryByText(/Loading/)).toBeNull();

    expect(screen.getByText("React")).toBeInTheDocument();
    expect(screen.getByText("Redux")).toBeInTheDocument();
    // Item does not have a button with Dismiss
    // expect(screen.getAllByText("Dismiss").length).toBe(2);
  });

  // App is not re-rendering for some reason
  // The fail on fetch runs but the html element does not change
  // test("fails fetching data", async () => {
  //   const promise: Promise<any> = Promise.reject();
  //   mockedAxios.get.mockImplementationOnce(() => promise);

  //   render(<App />);

  //   expect(screen.getByText(/Loading/)).toBeInTheDocument();

  //   try {
  //     await act(() => promise);
  //   } catch (err) {}
  //   expect(screen.queryByText(/Loading/)).toBeNull();
  //   expect(screen.getByText(/went wrong/)).toBeInTheDocument();
  // });

  test("removes a story", async () => {
    const promise: Promise<any> = Promise.resolve({
      data: {
        hits: stories,
      },
    });

    mockedAxios.get.mockImplementationOnce(() => promise);

    render(<App />);

    await act(() => promise);

    expect(screen.getAllByAltText("Dismiss").length).toBe(2);
    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();

    fireEvent.click(screen.getAllByAltText("Dismiss")[0]);

    expect(screen.getAllByAltText("Dismiss").length).toBe(1);
    expect(screen.queryByText("Jordan Walke")).toBeNull();
  });

  test("searches for specific stories", async () => {
    const reactPromise: Promise<any> = Promise.resolve({
      data: {
        hits: stories,
      },
    });
    const anotherStory = {
      title: "JavaScript",
      url: "https://en.wikipedia.org/wiki/JavaScript",
      author: "Brendan Eich",
      num_comments: 15,
      points: 10,
      objectID: 3,
    };
    const javascriptPromise: Promise<any> = Promise.resolve({
      data: {
        hits: [anotherStory],
      },
    });
    mockedAxios.get.mockImplementation((url) => {
      if (url.includes("React")) {
        return reactPromise;
      }

      if (url.includes("JavaScript")) {
        return javascriptPromise;
      }
      throw Error();
    });

    // Initial Render
    render(<App />);
    // First Data Fetching
    await act(() => reactPromise);
    expect(screen.getByDisplayValue("React")).toBeInTheDocument();
    expect(screen.queryByDisplayValue("JavaScript")).toBeNull();
    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();

    expect(screen.getByText("Dan Abramov, Andrew Clark")).toBeInTheDocument();
    expect(screen.queryByText("Brendan Eich")).toBeNull();
  });
});
