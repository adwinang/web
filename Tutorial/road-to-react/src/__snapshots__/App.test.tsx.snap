// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders snapshot 1`] = `
<form
  className="searchForm"
  onSubmit={[MockFunction]}
>
  <label
    className="label"
    htmlFor="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    className="input"
    id="search"
    onChange={[MockFunction]}
    type="text"
    value="React"
  />
  <button
    className="button buttonLarge"
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</form>
`;
